src/hotspot/share/runtime/vmStructs.hpp

```diff
@@ -188,13 +188,18 @@ class VMStructs {
 #ifdef ASSERT
 
 // This macro checks the type of a VMStructEntry by comparing pointer types
-#define CHECK_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type)                 \
- {typeName *dummyObj = NULL; type* dummy = &dummyObj->fieldName;                   \
-  assert(offset_of(typeName, fieldName) < sizeof(typeName), "Illegal nonstatic struct entry, field offset too large"); }
+#define CHECK_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type) {               \
+    char space[sizeof (typeName)];                                                 \
+    typeName *dummyObj = (typeName *)space; type* dummy = &dummyObj->fieldName;    \
+    assert(offset_of(typeName, fieldName) < sizeof(typeName), "Illegal nonstatic struct entry, field offset too large"); \
+  }
 
 // This macro checks the type of a volatile VMStructEntry by comparing pointer types
-#define CHECK_VOLATILE_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type)        \
- {typedef type dummyvtype; typeName *dummyObj = NULL; volatile dummyvtype* dummy = &dummyObj->fieldName; }
+#define CHECK_VOLATILE_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type) {      \
+    char space[sizeof (typeName)];                                                 \
+    typedef type dummyvtype; typeName *dummyObj = (typeName *)space;               \
+    volatile dummyvtype* dummy = &dummyObj->fieldName;                             \
+  }
```

Comment:
I think a better definition for this macro is
```
#define CHECK_VOLATILE_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type) \
  CHECK_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, std::add_volatile_t<type>)
```
(and `#include <type_traits>`)

Reply:
The member value type checking could be written more explicitly (and IMO more
clearly, though metaprogramming :) ) as:

```
#define CHECK_NONSTATIC_VM_STRUCT_ENTRY(typeName, fieldName, type) { \
  std::static_assert( \
    std::is_convertible< \
      std::add_pointer_t<decltype(std::declval<typeName>().fieldName)>, \
      std::add_pointer_t<type>>::value, \
    "type mismatch for " XSTR(fieldName) " member of " XSTR(typeName)); \
  assert(offset_of(typeName, fieldName) < sizeof(typeName), "..."); \
}
```

except that uses `std::declval`, which is in `<utility>`, which is not (yet)
permitted for use in HotSpot code.  Maybe that will change someday.  In the
meantime, we could add our own definition to globalUtilities.hpp, since it is
quite trivial.

```
template<typename T>
std::add_rvalue_reference_t<T> declval() noexcept;
```

I tried using `std::is_same` instead of `std::is_convertible`, which seems
like it might be better. However, there were a number of failures because some
uses are not exact matches. There were differences in volatile, in base vs
derived types, and reference vs non-reference types (if you drop the
conversion to pointers).

This dodges the whole question of alignment of the space buffer.